File: Blood_Donation-teddy-main/Frontend/lib/src/features/core/models/blood_request_model.dart
List<BloodRequest> bloodRequestFromJson(dynamic str) =>
    List<BloodRequest>.from((str).map((x) => BloodRequest.fromJson(x)));

class BloodRequest {
  String bloodType;
  String reason;
  double unitRequired;
  String deadLine;
  String hospital;
  String? personInCharge;
  String contactNumber;
  String patientName;
  String location;
  String? requestId;
  List<String>? userDoners;
  int? pendingState;
  int? completedState;

  BloodRequest({
    required this.bloodType,
    required this.reason,
    required this.unitRequired,
    required this.deadLine,
    required this.hospital,
    this.personInCharge,
    required this.contactNumber,
    required this.patientName,
    required this.location,
    this.requestId,
    this.userDoners,
    this.pendingState,
    this.completedState,
  });

  factory BloodRequest.fromJson(Map<String, dynamic> json) {
    return BloodRequest(
      requestId: json['requestId'],
      bloodType: json['bloodType'],
      reason: json['reason'],
      unitRequired: json['unitRequired'],
      deadLine: json['deadLine'],
      hospital: json['hospital'],
      personInCharge: json['personInCharge'],
      contactNumber: json['contactNumber'],
      patientName: json['patientName'],
      location: json['location'],
      userDoners: json['userDoners'] != null
          ? List<String>.from(json['userDoners'].map((x) => x.toString()))
          : null,
      pendingState: json['pendingState'],
      completedState: json['completedState'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bloodType': bloodType,
      'reason': reason,
      'unitRequired': unitRequired,
      'deadLine': deadLine,
      'hospital': hospital,
      'personInCharge': personInCharge,
      'contactNumber': contactNumber,
      'patientName': patientName,
      'location': location,
      'requestId': requestId,
    };
  }
}

File: Blood_Donation-teddy-main/backend/src/models/blood_request.model.ts
import mongoose, { Model, Schema, Document, Types } from "mongoose";
import { IUserDoner } from "./donation.model.js";

interface IBloodRequest {
  bloodType: string;
  reason: string;
  unitRequired: number;
  deadLine: string;
  hospital: string;
  personInCharge: string;
  contactNumber: string;
  patientName: string;
  location: string;
  requesterId?: string | Types.ObjectId;
  pendingState?: number;
  completedState?: number;
}

interface IBloodRequestDocument extends IBloodRequest, Document {
  requestId: string;
  userDoners: IUserDoner[];
}

const bloodRequestSchema = new Schema<IBloodRequestDocument>(
  {
    bloodType: { type: String, required: true },
    reason: { type: String, required: true },
    unitRequired: { type: Number, required: true },
    deadLine: { type: String, required: true },
    hospital: { type: String, required: true },
    personInCharge: { type: String },
    contactNumber: { type: String },
    patientName: { type: String },
    location: { type: String },
    requesterId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    userDoners: [{ type: Types.ObjectId, ref: "UserDoner" }],
    pendingState: { type: Number, default: 0 },
    completedState: { type: Number, default: 0 },
  },
  {
    timestamps: true,
    toJSON: {
      transform: (_doc, ret) => {
        ret.requestId = ret._id.toString();
        delete ret._id;
        delete ret.__v;
      },
    },
  }
);

const BloodRequest: Model<IBloodRequestDocument> =
  mongoose.model<IBloodRequestDocument>("BloodRequest", bloodRequestSchema);

export { IBloodRequest, IBloodRequestDocument, BloodRequest };
